### ISSUE TITLE
Develop proposal to establish universal API for backends

### ISSUE THREAD

**Issue by gubuntu** [kartoza/geonode#144]
The goal here will be to provide a __proposal__ for a standardised, formalised abstraction layer between the GeoNode platform and the underlying data store backend. This abstraction layer will provide a way for new backends to be easily integrated since they need only implement a few well described API calls and not need to make extensive changes to the underlying GeoNode code base. Examples of key API entry points will be:
- Rendering layer tiles
- Obtaining OGC capabilities for a layer
- Map composition
- Print requests
- Registering a new layer
- Deleting a layer
- Registering a layer style
- Deleting a layer style 
- Retrieving a list of layer styles 
- And so on...

Implementing this would imply moving any Geoserver or QGIS specific code from core GeoNode to their respective backend apps

**Comment by gubuntu** [comment]
Extend the GeoNode API docs to:
1. expose functionality available in the QGIS server backend that is not available in Geoserver
1. for each API component, identify clearly which backends are supported so they can be filtered based on which backend is configured. 
2. meet requirements of QGIS desktop and GeoNode frontend clients, even if the functionality it not available yet. Identify these as future work so they can also be filtered out of a production instance but are still available to develop against.

**Comment by gubuntu** [comment]
we are already using a Python decorator for tests to automatically switch between beackend-specific code. This could be applied in production.

**Comment by lucernae** [comment]
After reviewing GeoNode's code I found out several issues and limitations that we have to deal with.
Along the progress, we managed to patch things to make it properly work and have robusts unittests implemented for both GeoServer and QGIS Server. But this is still not done in a structured way that makes it easier for new developer to understand how GeoNode and OGC backend communicates. At the end we come up with several proposals on how to improve the structure.

# Challenges and Limitations

As expected, we have several challenges because GeoNode was originally developed using only GeoServer backend in mind. There is no alternative in original design. With this we encountered general issues:

- The codes were developed tightly with GeoServer functionality in mind. In some packages, like security, there are too many functions that assumes that it will communicate with GeoServer. It makes it difficult and ugly to put QGIS Server related codes in a function that the sole purpose is to communicate with GeoServer.
- It is not straightforward to see what GeoNode expects the backend to do from the code. Some codes directly ask GeoServer without a wrapper.
- It is not clear what GeoNode can provides to the map frontend. It becomes especially difficult because Frontend code (HTML + Javascript) doesn't have standardized code structure by default.

# The way we solve it

At the moment, we already implement a small degree of abstractions to let GeoNode knows which codes must be executed for particular backends. This improves code readability and structures, and helps to run unittests in parallel.

- The use of Python decorator to skip function executes if it is intended for different backend:

For example:
https://github.com/kartoza/geonode/blob/1d8aeeadeed9343d6b71b1a8bc0cabd329b73b27/geonode/base/models.py#L1053

We use python decorator to tag python function that is specific for GeoServer backend. This way, we can easily knows that this function will not be executed if QGIS Server is used.

- The use of simplified backend type check

In the past, we used ```if `geonode.geoserver` in settings.INSTALLED_APPS```. This is too long, not elegant, and doesn't convey the meaning that a particular block is executed on particular backend. We change this slightly into ```if check_ogc_backend(qgis_server.BACKEND_PACKAGE):```. With this, we know easily that this code block will be executed if it is using QGIS Server backend, and the settings were configured correctly.

- Implement unittests for QGIS Server backend. We have to somehow make sure that future changes will not regress the codes. We implemented unittests in the qgis_server package itself, some other places in different packages (integration tests) and also make sure existing generic tests were also passed for QGIS Server backend.

- The use of Travis to execute unittests in parallel for each backend.

Previously, GeoNode uses Travis to run the unittests, but we didn't have unittests for QGIS Server. We improved the setup so Travis will create 2 job/environment that each run the same tests suite for each backend. Of course, some of the unittests were skipped if it is too specific for each backend. We skip it using the same python decorator we explained previously.
By running the job in parallel, we make sure that unittests were passed for each backend. The tests itself run more or less the same time.

<img width="1065" alt="screen shot 2017-09-22 at 16 15 25" src="https://user-images.githubusercontent.com/831865/30737576-559f7696-9fb1-11e7-92a1-158be4952641.png">

- Switch to use codecov (previously used coveralls)

The reason to switch is to have aggregated code coverage. By executing unittests for GeoServer and QGIS Server, we will have 2 code coverage patterns, with each missing a coverage from the other one. Codecov is able to aggregate both coverage, so we can know the overall coverage percentage of the code. This is not possible in coveralls.

# Proposed Improvements

These changes described previously were made with intention to introduce as little changes as possible to upstream GeoNode. However to be able to support multiple backends extensively (and clearly). We need to have architectural design in the code structure. There are several ideas come to mind:

- Create a more layered architecture by adding another wrapper for generic backend.

Adding wrapper layer/package will introduce a common functionalities that GeoNode expects for the backend to be supported. This will introduce the same basic interface that the backend will support.

- Introduce an abstract module level API to be used internally by GeoNode, instead of directly calling backend code.

The code needs to be changed to call an abstracted function/class of the backend. This way the code organization will be neat, and backend code will call an abstracted function/class of GeoNode or vice versa. Some example will be to introduce an abstract class or modules of OGC backend with predefined sets of methods and function that GeoNode expects to implement. Then each backend implemented by inherit from the base class or implement the same function signatures. This level of abstractions should happen across the modules that potentially have OGC backend implementation details. For example, Layer, Maps, Security modules should have generic implementations and let each backend implement additional logic, like supported file type or WMS links.

- Introduce a predefined sets of abstract functionality of views that the map frontend library should support.

At the moment, it is hard to figure out what kind of functionality these different frontend library (like leaflet, openlayers, GeoExt) will have in common and what kind of views/javascript function needs to be supported.

**Comment by gubuntu** [comment]
thanks @lucernae, I've incorporated this into the report
